chalk-ir/src/lib.rs

```diff
@@ -836,6 +836,17 @@ impl<I: Interner> AliasTy<I> {
     pub fn intern(self, interner: &I) -> Ty<I> {
         Ty::new(interner, self)
     }
+
+    pub fn type_parameters<'a>(&'a self, interner: &'a I) -> impl Iterator<Item = Ty<I>> + 'a {
```

Comment:
Is this only used to support `self_type_parameter` below? It's a bit of a dubious thing to request otherwise, so maybe we should remove it if it is dead. (It encourages you to ignore lifetimes/constants, we've seen in rustc that's that's a bit error-prone.)